proposals/p2360.md

```diff
@@ -0,0 +1,478 @@
+    but it could still implicitly convert to `type`, as `()` does, if we so
+    desire.
+    -   This extends to other cases: we could say that for integer literals,
+        which we have recently been describing as having type `IntLiteral(n)`,
+        there is an `impl IntLiteral(n) as ImplicitAs(type)` which converts a
+        value of type `IntLiteral(n)` to the `type` value `IntLiteral(n)`. This
+        would then permit `let v: 0 = 0;`.
+    -   Such changes to `i32` or integer literals are not part of this proposal.
+
+## Details
+
+### Terminology
+
+The changes to the language rules in this proposal are fairly small, but there
+is a significant change in how we describe those rules. We now use the following
+terminology:
+
+-   A _type_ is a value of type `type`.
+-   A _facet type_ is a constrained or unconstrained version of `type`, such as
+    an interface type or a named constraint type. We previously referred to a
+    facet type as a type-of-type, but that is no longer accurate, as the type of
+    any type is now by definition `type`.
+-   A _facet_ is a value of a facet type. For example, `i32 as Hashable` is a
+    facet, and `Hashable` is a facet type. Note that all types are facets.
+    -   A facet value can be thought of as a tuple of a possibly-symbolic
+        identity for a type, plus any witnesses needed to demonstrate that the
+        type satisfies the constraints of the facet type. A type is then a facet
+        value that has no witnesses.
+-   We use the term _generic type_ to refer to a type or facet introduced by a
+    `:!` binding, such as in a generic parameter or associated constant.
+    -   This gives rise to terms such as _generic type parameter_ and
+        _associated generic type_.
+    -   It is worth noting that a generic type is not necessarily a type, as its
+        type may be an interface or constraint rather than `type`.
+
+### `type` and `Core.Type`
+
+This proposal introduces the name `type` as a keyword -- a type literal -- that
+names the facet type with an empty set of constraints. This keyword replaces the
+provisional name `Type` for this functionality.
+
+In [issue #2113](https://github.com/carbon-language/carbon-lang/issues/2113), it
+was decided that such functionality may be an alias for a name in the prelude,
+such as `Core.Type`. This proposal does not introduce a corresponding prelude
+name. The facet type `type` is primitive and provided by the compiler, and not
+an alias.
+
+[As noted in the generics design](/docs/design/generics/details.md#named-constraints),
+a declaration such as `constraint MyVersionOfType {}` would introduce a facet
+type that is equivalent to `type`. This proposal does not change this decision.
+Because such a named constraint `MyVersionOfType` is equivalent to `type`, that
+is, because they are the same facet type, a value `T:! MyVersionOfType` is a
+type under this proposal.
```

Comment:
I'm comfortable with this, but I think it would be good to include a rationale for the decision here?

And/or to include using an empty named constraint in the prelude in the alternatives considered?

Not looking for much, just to record a key point or two from the discussion we had here.

Reply:
My intent is that this proposal is not making a decision here. I think we might want a different outcome in this case. I've opened an issue for leads #2409 to discuss this further.